src/authz-smart/smartHandler.ts

```diff
@@ -8,31 +8,136 @@ import {
     AuthorizationRequest,
     AuthorizationBundleRequest,
     AllowedResourceTypesForOperationRequest,
+    SystemOperation,
+    TypeOperation,
 } from 'fhir-works-on-aws-interface';
-import { SMARTConfig } from './smartConfig';
+import axios from 'axios';
+import { LaunchType, ScopeType, SMARTConfig } from './smartConfig';
 
 // eslint-disable-next-line import/prefer-default-export
 export class SMARTHandler implements Authorization {
+    static readonly CLINICAL_SCOPE_REGEX = /^(patient|user|system)\/([a-zA-Z]+|\*)\.(read|write|\*)$/;
+
+    static readonly LAUNCH_SCOPE_REGEX = /^(launch)(\/(patient|encounter))?$/;
+
     private readonly version: number = 1.0;
 
     private readonly config: SMARTConfig;
 
     constructor(config: SMARTConfig) {
-        this.config = config;
-        if (this.config.version !== this.version) {
-            throw Error('Configuration version does not match handler version');
+        if (config.version !== this.version) {
+            throw Error('Authorization configuration version does not match handler version');
         }
+        this.config = config;
     }
 
     async isAuthorized(request: AuthorizationRequest): Promise<boolean> {
+        const authZPromise = axios.post(
+            this.config.authZUserInfoUrl,
+            {},
+            { headers: { Authorization: `Bearer ${request.accessToken}` } },
+        );
+
+        const decoded = decode(request.accessToken, { json: true }) || {};
+        const { aud, iss } = decoded;
+        // verify aud & iss
+        if (this.config.expectedAudValue !== aud || this.config.expectedIssValue !== iss) {
+            console.error('aud or iss is not matching');
+            return false;
+        }

```

Comment:
I *think* that the authZPromise promise can crash the server since it ends up as an unhandled rejection if this function returns false early.

Regarding the code order, after the discussion I think that it was better as it was originally written. I guess it's fine to check some claims like aud first so that some bad tokens are rejected early without needing the call to `/userinfo`.

Maybe just add a comment above the call to jwt.decode mentioning that the token will be validated by the /userInfo call. It's just that using `jwt.decode` instead of `jwt.verify` is usually a massive red flag that can confuse people reading this code, but here the overall implementation is correct.

Reply:
interesting when running the unit tests I didn't see those sort of failures but good callout it may be one